Add vDPA support

sboeuf opened the pull request:
Adding the vDPA support through this PR.
vIOMMU and migration support will come later.

sboeuf commented:
@rbradford PR's ready, but having Hirsute Hippo workers is preventing the CI from passing. We need a 5.12 kernel on the host, which we have with Focal and Impish (5.13 for both), but since Hirsute is deprecated, it's still using 5.11.

rbradford commented on scripts/prepare_vdpa.sh:
```diff
@@ -0,0 +1,39 @@
+sudo modprobe vdpa_sim
+sudo modprobe vdpa_sim_blk
+sudo modprobe vdpa_sim_net
+git clone https://github.com/shemminger/iproute2.git
```
If you can wait for us to move to jammy for the container the binary is already present: https://packages.ubuntu.com/search?suite=jammy&arch=any&mode=exactfilename&searchon=contents&keywords=vdpa

sboeuf replied:
Well it's up to you, but I can remove it later when we move to Jammy. I mean the download/build/install of `iproute2` is super fast so I'm not too worried.

rbradford replied:
Please open a GH issue to track this.

rbradford replied:
If you actually mention it in a github comment in the code I think it might cross-reference back but at the very least it is easy to search for.

rbradford replied:
Dittor

sboeuf replied:
Done!

rbradford reviewed (commented):
This approach breaks the ability to run `./scripts/dev_cli.sh tests --integration` outside of the Jenkins CI. Please can you look into solutions that would allow that to continue to work e.g. by possibly skipping those tests when not in the appropriate environment.

sboeuf commented:
> This approach breaks the ability to run `./scripts/dev_cli.sh tests --integration` outside of the Jenkins CI.

Yes it breaks this ability.

>Please can you look into solutions that would allow that to continue to work e.g. by possibly skipping those tests when not in the appropriate environment.

Of course, but are we looking at detecting the lack of support from the tests (I mean `scripts/run_integration_x86_64.sh` or `tests/integration.rs`), OR are we thinking about defining an extra parameter to `dev_cli.sh` that would pass an environment variable down to the tests in order to know what to do?

rbradford commented:
> > This approach breaks the ability to run `./scripts/dev_cli.sh tests --integration` outside of the Jenkins CI.
> 
> Yes it breaks this ability.
> 
> > Please can you look into solutions that would allow that to continue to work e.g. by possibly skipping those tests when not in the appropriate environment.
> 
> Of course, but are we looking at detecting the lack of support from the tests (I mean `scripts/run_integration_x86_64.sh` or `tests/integration.rs`), OR are we thinking about defining an extra parameter to `dev_cli.sh` that would pass an environment variable down to the tests in order to know what to do?

Or it could even be in the test to check to see if the module is loaded.

sboeuf commented:
> > > This approach breaks the ability to run `./scripts/dev_cli.sh tests --integration` outside of the Jenkins CI.
> > 
> > 
> > Yes it breaks this ability.
> > > Please can you look into solutions that would allow that to continue to work e.g. by possibly skipping those tests when not in the appropriate environment.
> > 
> > 
> > Of course, but are we looking at detecting the lack of support from the tests (I mean `scripts/run_integration_x86_64.sh` or `tests/integration.rs`), OR are we thinking about defining an extra parameter to `dev_cli.sh` that would pass an environment variable down to the tests in order to know what to do?
> 
> Or it could even be in the test to check to see if the module is loaded.

Yep, that's what I meant by `tests/integration.rs`.

sboeuf commented on tests/integration.rs:
```diff
@@ -5877,6 +5877,206 @@ mod parallel {
+    #[cfg(target_arch = "x86_64")]
+    fn test_vdpa_block() {
+        // Before trying to run the test, verify the vdpa_sim_blk module is correctly loaded.
+        if !exec_host_command_status("lsmod | grep vdpa_sim_blk").success() {
+            return;
+        }
```
@rbradford Here's the check

sboeuf commented on tests/integration.rs:
```diff
@@ -5877,6 +5877,206 @@ mod parallel {
+    #[test]
+    #[cfg(target_arch = "x86_64")]
+    fn test_vdpa_net() {
+        // Before trying to run the test, verify the vdpa_sim_net module is correctly loaded.
+        if !exec_host_command_status("lsmod | grep vdpa_sim_net").success() {
+            return;
+        }
```
Ditto

likebreath reviewed (approved):
Changes look good to me. Helped me learned/refreshed many things around device-manager crate and the vhost crate.

I wonder do we see some performance gains on block/net by using vDPA?

sboeuf commented:
@likebreath 
>I wonder do we see some performance gains on block/net by using vDPA?

If they were backed by HW devices (like Intel or Mellanox NICs) you'd see a huge performance gain. Unfortunately we don't have access to such devices at the moment, meaning the tests are performed with the in kernel simulators.

rbradford commented on virtio-devices/src/vdpa.rs:
```diff
@@ -0,0 +1,374 @@
+    #[error("Failed to get queue size: {0}")]
+    GetVringNum(vhost::Error),
+    #[error("Invalid IOVA range")]
+    InvalidIovaRange,
```
Include the range in then enum to ease debugging?

rbradford commented on virtio-devices/src/vdpa.rs:
```diff
@@ -0,0 +1,374 @@
+        // to fix it. Below is the correct code once the bug will be fixed:
+        //
+        // let iova_range = vhost.get_iova_range().map_err(Error::GetIovaRange)?;
+        let iova_range = VhostVdpaIovaRange {
```
For situations like this I would like us to have a bug open in our issue tracker which we can then use to track this.

sboeuf replied:
#3861

rbradford commented on virtio-devices/src/vdpa.rs:
```diff
@@ -0,0 +1,374 @@
+
+    fn read_config(&self, offset: u64, data: &mut [u8]) {
+        if let Err(e) = self.vhost.get_config(offset as u32, data) {
+            error!("vdpa: Failed reading virtio config: {}", e);
```
`error!()` include source file and line so i'm not sure the extra "vdpa" prefix adds anything?

rbradford commented on virtio-devices/Cargo.toml:
```diff
@@ -26,9 +26,10 @@ seccompiler = "0.2.0"
 serde = "1.0.136"
```
Maybe expand on the commit message a tad? Just explain what a VDPA device is at a high level.

rbradford commented on vmm/src/api/openapi/cloud-hypervisor.yaml:
```diff
@@ -921,6 +925,26 @@ components:
+        num_queues:
+          type: integer
+          default: 1
+        iommu:
```
I think you didn't implement the iommu support yet so would be good to leave this out for now?

rbradford commented on vmm/src/config.rs:
```diff
@@ -1851,6 +1860,77 @@ impl UserDeviceConfig {
+    #[serde(default = "default_vdpaconfig_num_queues")]
+    pub num_queues: usize,
+    #[serde(default)]
+    pub iommu: bool,
```
Ditto.

rbradford commented on vmm/src/config.rs:
```diff
@@ -2374,6 +2455,19 @@ impl VmConfig {
+            let mut vdpa_config_list = Vec::new();
+            for item in vdpa_list.iter() {
+                let vdpa_config = VdpaConfig::parse(item)?;
+                if vdpa_config.iommu {
```
Ditto.

rbradford commented on vmm/src/seccomp_filters.rs:
```diff
@@ -583,6 +628,15 @@ fn create_vcpu_ioctl_seccomp_rule() -> Result<Vec<SeccompRule>, BackendError> {
         and![Cond::new(1, ArgLen::Dword, Eq, VFIO_GROUP_UNSET_CONTAINER)?],
         and![Cond::new(1, ArgLen::Dword, Eq, VFIO_IOMMU_UNMAP_DMA)?],
+        and![Cond::new(1, ArgLen::Dword, Eq, VHOST_VDPA_SET_STATUS)?],
+        and![Cond::new(1, ArgLen::Dword, Eq, VHOST_VDPA_GET_CONFIG)?],
```
Do we try to keep in this in lexical order?

sboeuf replied:
No I was following the order in the vhost header: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/uapi/linux/vhost.h

rbradford commented on src/bin/ch-remote.rs:
```diff
@@ -486,6 +508,15 @@ fn main() {
         )
+        .subcommand(
+            Command::new("add-vdpa")
+                .about("Add persistent memory device")
```
Copy and paste error!

rbradford commented on tests/integration.rs:
```diff
@@ -5877,6 +5877,206 @@ mod parallel {
+    fn test_vdpa_block() {
+        // Before trying to run the test, verify the vdpa_sim_blk module is correctly loaded.
+        if !exec_host_command_status("lsmod | grep vdpa_sim_blk").success() {
+            return;
```
I was going to say it might be worth printing a message to say you're skipping it. But stderr is only shown on test failure. It's a shame `cargo test` lacks a skip ability.

sboeuf replied:
Yeah that's too bad :(

rbradford reviewed (commented):
Very impressed with how quickly you got this feature off the ground.

sboeuf commented:
@rbradford 
> Very impressed with how quickly you got this feature off the ground.

Thanks :)

I think I addressed all your comments, so please have a look at the new version.

rbradford reviewed (approved):
Does this need a new file in `docs/` ?
